docs/architecture/adr-002-ibc-relayer.md

```diff
@@ -0,0 +1,644 @@
+- the IBC modules on chain correctly implement and respond to queries
+  - [IBC-Modules-Rust] an implementation for some queries currently exist in Cosmos-SDK and same and more need to be implemented in Rust. The full requirements are detailed in section Relayer Queries. 
+- the relayer needs the ability to send rpc/http ABCI queries to and receive replies from Tendermint/Cosmos-SDK
+  - [[ABCI Rust](https://github.com/tendermint/rust-abci)] - ABCI Rust implementation
+  - [IBC-Modules-Rust] identifier validation is required (ICS-024)
+  - [IBC-Modules-Rust] requires Rust types for all query responses
+  - [[Merkle-Proofs-Rust](https://github.com/confio/ics23/tree/master/rust)] (candidate implementation) - some query responses include proofs and included in IBC transactions by the relayer (some may be validated, TBD)
+
+#### IBC Messages
+The relayer creates transactions that include IBC messages to manage clients, connections and channels, and send application packets to destination chains. These messages must be defined in the IBC Rust implementation [IBC-Modules-Rust].
+
+#### IBC Logging System
+IBC packet data & timeouts are not stored directly in the chain state (as this storage is presumed to be expensive) but are instead committed to with a succinct cryptographic commitment (only the commitment is stored).
+As a consequence, IBC requires that a **host state machine MUST provide an event logging system** that logs data in the course of transaction execution. **Logs must be queryable** by relayers to read IBC packet data & timeouts. 
+
+The logging system must provide the following functions:
+ - [IBC-Modules-Go] emitLogEntry for emitting log entries called by the state machine during transaction execution:
+   - type emitLogEntry = (topic: string , data: []byte) => void
+   - example: emitLogEntry("sendPacket", {sequence: packet.sequence , data: packet.data, timeout: packet.timeout})
+ - [IBC-Modules-Go] queryByTopic for querying past logs matching a given topic:
+   - type queryByTopic = (height: uint64 , topic: string) => Array < []byte >
+
+#### Keyring 
+The relay process must have access to its accounts with tokens on all destination chains, with sufficient balance to pay for transaction fees. Account key information must be stored and managed securely. A keyring implementation is required for CRUD key operations. 
+[Keyring-Rust] Investigation in existing Rust implementations is needed. (ex: [hwchen-keyring](https://github.com/hwchen/keyring-rs))
+
+### Chain Transactions and Signing
+The relayer must create chain specific signed transactions. 
+[Cosmos-Tx-Rust] For the first release Cosmos-SDK transaction signing is required. One possible implementation is [iqlusion's sdtx crate](https://github.com/iqlusioninc/crates/tree/develop/stdtx)
+
+#### Implementation of IBC "routing module"
+The default IBC handler uses a receiver call pattern, where modules must individually call the IBC handler in order to bind to
```

Comment:
This part is not very clear.

Reply:
Had some issues with this as well, until I read [this](https://github.com/cosmos/ics/blob/a4173c91560567bdb7cc9abee8e61256fc3725e9/ibc/5_IBC_DESIGN_PATTERNS.md#call-receiver).